Apps/Inc/fifo_protected.h

```diff
@@ -0,0 +1,337 @@
+ * @return true if successful
+ * @return false if unsuccessful
+ */
+static bool __attribute__((unused))
+GET (FIFO_TYPE_NAME *fifo, FIFO_TYPE *elem) {
+    OS_ERR err;
+    CPU_TS ticks;
+    bool success = false;
+    
+    OSSemPend(&(fifo->FIFO_SEM4_NAME), 0, OS_OPT_PEND_BLOCKING, &ticks, &err);
+    assertOSError(OS_NONE_LOC, err);
+
+    OSMutexPend(&(fifo->FIFO_MUTEX_NAME), 0, OS_OPT_PEND_BLOCKING, &ticks, &err);
+    assertOSError(OS_NONE_LOC, err);
+    
+    if(!IS_EMPTY(fifo)) {
+        *elem = fifo->buffer[fifo->get];
+        fifo->get = (fifo->get + 1) % FIFO_SIZE;
+        success = true;
+    }
+    
+    OSMutexPost(&(fifo->FIFO_MUTEX_NAME), OS_OPT_POST_NONE, &err);
+    assertOSError(OS_NONE_LOC, err);
+
+    return success;
+}
+
+/**
+ * @brief Put an element into the fifo
+ * 
+ * If the type of the fifo is myfifo_t, then this function
+ * will be called myfifo_put().
+ * 
+ * @param fifo A pointer to the fifo
+ * @param elem The element to put
+ * @return true if successful
+ * @return false if unsuccessful
+ */
+static bool __attribute__((unused))
+PUT (FIFO_TYPE_NAME *fifo, FIFO_TYPE elem) {
+    OS_ERR err;
+    CPU_TS ticks;
+
+    bool success = false;
+
+    OSMutexPend(&(fifo->FIFO_MUTEX_NAME), 0, OS_OPT_PEND_BLOCKING, &ticks, &err);
+    assertOSError(OS_NONE_LOC, err);
+    
+    if(!IS_FULL(fifo)) {
+        fifo->buffer[fifo->put] = elem;
+        fifo->put = (fifo->put + 1) % FIFO_SIZE;
+        success = true;
+    }
+
+    OSMutexPost(&(fifo->FIFO_MUTEX_NAME), OS_OPT_POST_NONE, &err);
+    assertOSError(OS_NONE_LOC, err);
+
+    if(success){
+        OSSemPost(&(fifo->FIFO_SEM4_NAME), OS_OPT_POST_1, &err);
+        assertOSError(OS_NONE_LOC, err);
+    }
+
+    return success;
+}
+
+/**
+ * @brief Peek into the fifo
+ * 
+ * If the type of the fifo is myfifo_t, then this function
+ * will be called myfifo_peek().
+ * 
+ * @param fifo A pointer to the fifo
+ * @param elem A pointer to space where the next element will be put
+ * @return true if successful
+ * @return false if unsuccessful
+ */
+static bool __attribute__((unused))
+PEEK (FIFO_TYPE_NAME *fifo, FIFO_TYPE *elem) {
+    OS_ERR err;
+    CPU_TS ticks;
+
+    bool success = false;
+
+    OSMutexPend(&(fifo->FIFO_MUTEX_NAME), 0, OS_OPT_PEND_BLOCKING, &ticks, &err);
+    assertOSError(OS_NONE_LOC, err);
+
+    if(!IS_EMPTY(fifo)) {
+        *elem = fifo->buffer[fifo->get];
+        success = true;
+    }
+
+    OSMutexPost(&(fifo->FIFO_MUTEX_NAME), OS_OPT_POST_NONE, &err);
+    assertOSError(OS_NONE_LOC, err);
+
+    return success;
+}
+
+/**
+ * @brief Take the last element of the fifo (most recent)
+ * 
+ * If the type of the fifo is myfifo_t, then this function
+ * will be called myfifo_popback().
+ * 
+ * @param fifo A pointer to the fifo
+ * @param elem A pointer to space to put the element
+ * @return true if successful
+ * @return false if unsuccessful
+ */
+static bool __attribute__((unused))
+POPBACK (FIFO_TYPE_NAME *fifo, FIFO_TYPE *elem) {
```

Comment:
Just a prefrence, but I'd like it if it was called just POP instead.

Reply:
I'm with Nathaniel, popback on a fifo sounds like you're popping the back of it which would be the most recent element (ie. stack type behavior). Conceivably the fifo could be generalized into a deque which could make this more important but thats out of the scope of this pr.

Reply:
Popback makes sense to me, but back and front are confusing when it comes to fifos and stacks. That's why you'll see a lot of fifo implementations call them get and put instead.

Reply:
Popback was the original name in fifo.h 🤣 but I can change it in both if y'all want.

Reply:
Maybe just conform to the C++ std queue api def